### ISSUE TITLE
saving username/password login per server basis

### ISSUE THREAD

**Issue by sfan5** [minetest/minetest#1566]
```
<Mikeonline> sfan5: feature request: saving username/password login per server basis
<Mikeonline> and a dropdownbox for saved usernames :)
```

**Comment by lightonflux** [comment]
How could MT realise half-automatic login without storing the password or password hash? 

Something like a key exchange or certificates?

Example Process:
1. User registers on server with password
2. server saves password hash
3. Client generates hash of the password
4. Client encrypts a message with that hash and own public key
5. sends encrypted message to server
6. server stores public key
7. later the client uses the remember password function
8. the client sends a message (encrypted with server pub key) with the current time (epoch string) signed with its key to the server  
9. server checks if time is within 100sec (because of replay attacks) and if signed with the right key (that of the user) and if user is not logged in if not then discards the message 
10. login successful

Maybe not the most efficient and potential a target for DDoS. But it just a quick idea.

**Comment by rubenwardy** [comment]
I see this as quite a low priority feature

**Comment by paramat** [comment]
-1 for password due to security, and only a minor convenience for username. Doesn't seem worth the work.
:-1:

**Comment by paramat** [comment]
:-1: for passwords. Whenever i see the 'remember my password' option on websites i never use it because it is obviously less secure.

For usernames, this is a luxury nonessential feature, let's not bother.

**Comment by rubenwardy** [comment]
Password storage is more secure than manually typing a password if you use a random password with it, eg: `BgzxJYal4nEOqhKEMYbP`

However, I see this as low priority

**Comment by DS-Minetest** [comment]
Hm, saving user names would maybe be really useful for those who have names that are not accepted on some servers (eg. it's too short) and so have to choose another name on these servers but don't want to take this new name on the servers where they already have a player with their old name.

**Comment by rubenwardy** [comment]
> Whenever i see the super-lazy luxury 'remember my password' option on websites i never use it because it is obviously less secure.

For one, I've never seen an option like that. It's usually "remember me", which does *not* save your password. It saves a long-term session cookie rather than just one that expires quickly.

I support doing this, as it is good usability. It would be good to combine this will a nicer login-register system

**Comment by appgurueu** [comment]
I agree with rubenwardy. Users have already started to write [insecure "extensions"](https://forum.minetest.net/viewtopic.php?f=14&t=22313) to the mainmenu.

I would propose allowing some sort of "master key" which is used to hash your passwords, so you only have to remember a single password, and generating random & secure passwords for servers.

**Comment by paramat** [comment]
Sorry, i was wrong. Edited comment and removed disapproval. I still do not think it is worth doing but am not much bothered.

**Comment by sindastra** [comment]
I am in favor of a "remember me" feature.

There are two possible ways from the top of my head:

- Get a random token (session ID) from the server on successful login which then gets stored on the client and can expire or be revoked.
- Modern operating systems provide a key store to securely store and retrieve credentials. Integrating with one of those means offloading security to the OS so you don't have to deal with it.

The reason I am favoring this, is because I use long, random, unique passwords and having to unlock KeePass each time is tedious. I am sure some other users won't put up with this and instead use an insecure password.

Using an insecure password that could be guessed is in my opinion a greater security concern than trying to break into someone's machine and secure key store.

Besides, "remember me" would be optional.

**Comment by sindastra** [comment]
Here are some credential storage systems that can be interfaced by applications to store and request credentials.

macOS https://en.wikipedia.org/wiki/Keychain_(software)
KDE https://en.wikipedia.org/wiki/KWallet
GNOME https://en.wikipedia.org/wiki/GNOME_Keyring

These are encrypted and get decrypted on user login. If done right, an application can seamlessly store and retrieve its credentials.

Windows does also provide a credential storage system but I don't know enough about it!

**Comment by appgurueu** [comment]
> Here are some credential storage systems that can be interfaced by applications to store and request credentials.
> 
> macOS https://en.wikipedia.org/wiki/Keychain_(software)
> KDE https://en.wikipedia.org/wiki/KWallet
> GNOME https://en.wikipedia.org/wiki/GNOME_Keyring
> 
> These are encrypted and get decrypted on user login. If done right, an application can seamlessly store and retrieve its credentials.
> 
> Windows does also provide a credential storage system but I don't know enough about it!

I hate how this is extremely platform-specific, but it is what it is. We should go for an abstraction like https://github.com/hatstand/keychain though to simplify things.

**Comment by rubenwardy** [comment]
I have a branch with support for system keychains mostly implemented. The UX sucks though, having a dialog pop up is quite annoying

**Comment by mk-pmb** [comment]
Meanwhile I just use xdotool to click the password field, type the password, click login.

**Comment by mk-pmb** [comment]
Another idea: Can we hook a URL schema? Might also be nice for websites. A "Join our Server" link to `minetest://user:pass@server:port/`, confirm in browser, and ready you are.

**Comment by rubenwardy** [comment]
See #10373

**Comment by mk-pmb** [comment]
Thanks. I consider it relevant here, too, because it would offer a standardized cross-plattform login mechanism that all usual password managers should support.

**Comment by mk-pmb** [comment]
I discovered that MT 5.3.0 has all the command-line options required.
Just make a nice shell script (or desktop shortcut in Windows) where run the minetest client program with options like
`--address server23.mine.test --port 30000 --name "Guest" --password-file "X:\path\to\passwordfile.txt" --go`
The password file should be a plain text file that has your password and nothing else, not even a newline character at the end. Or you could use `--password "TopSecret"` which is less secure but a lot of players probably won't care. :>

**Comment by codeandfix** [comment]
storing an unencrypted password on a server is bad practice, storing it plain on client is actually not so bad, especially for games.
if i have to type my password everytime, then i tend to choose simple passwords which leads to very insecure passwords.
it is very terrible to enter a password manually for just a game... i like to be paranoid, but this paranoia here is far too much ;-)
anyway if you dont want to store a password, add a commandline option like mk-pmb told before so i can use shortcuts.
if you dont use your minetest password on other service, there is no risk if stolen from local client.
and if you are that afraid, take a look at traccars api key solution for login. here you create a key which is well known to server and client but randomly choosen by the server. so in other words every user will have 2 passwords.
one master choosen by the player and one by the server. the serverside-user-password could be used inside client to reconnect.

but we need a very quick fix, like adding cmdline or simple store first.
its the most terrible thing on minetest ever, that passwords are not stored ;)

**Comment by sindastra** [comment]
Or you could do what every website does: Store a session key. That way no password is stored and if the session key gets compromised, it would only be for the specific server(s).

Of course that would require support from both the client and the server, but it's better than storing a password.

**Comment by mk-pmb** [comment]
@sindastra Which party would check and authorize the session at login, and using what mechanism? Usually for websites, a hash of the password is stored, and clients need to send the actual password at login. Transport encryption is used to cover up the most obvious problems of this broken design. The broken-ness is so bad that in recent years, some browsers have even made attempts to shove proper crypto-auth onto end users.

**Comment by sfan5** [comment]
FYI Minetest has never stored plaintext passwords on the server and since 0.4.13 uses [SRP](https://en.wikipedia.org/wiki/Secure_Remote_Password_protocol), which provides all relevant security guarantees.

**Comment by codeandfix** [comment]
what i want to say is, that storing passwords on servers is bad, but storing plain text passwords on the users machine is not such a big problem. if a local machine is compromised, then normally the attacker has access to all the data of the user.
for servers the problem is, that here you have access to many users.

**Comment by mk-pmb** [comment]
> what i want to say is, that […] storing plain text passwords on the users machine is not such a big problem.

I agree. However, there are lots of password managers already, with lots of different security models. The desktop shortcut with a `--password` option is probably being the lowest.
For the task "login management" (password + username + server address/port), users already have lots of choices, so I see no need to add this task to our project scope. Rather, we should help users connect their login management software to MineTest. For me it seems that the easiest way to do so is via URL protocol handlers.

**Comment by codeandfix** [comment]
Using a password manager is a single point of failure. If the passwords are stored somewhere, like in games ini file or in shortcuts, it is much more secure if stored central. maybe a master password could protect these database, but a keylogger will get it, too.
so if someone is trying to steal your password, its normally some mod author which you cant trust much. thats why i prefer it inside the client. also the mod wont be supported for an endless time. if its inside the core client iam sure it will migrate forever.

i think adding command line parameters is an usual way for all games. so it should be supported. normally a player wants to connect to the same server everytime again and again, so i would click one shortcut and iam ingame in a few seconds.
another shortcut will open a test user. so i can have 2-3 users by one click.

also another nice idea would be identity faking for the admin to test protection.
i dont want to reconnect, just type /su foobar and then iam acting with his privs.

**Comment by mk-pmb** [comment]
> Using a password manager is a single point of failure.

That problem is independent from MineTest, and there are solutions for it. For most people's everyday logins it doesn't matter because password reset via email is a much weaker single point of failure.

Feel free to implement a multi-factor authentification access control mod and use it on your server.


> i think adding command line parameters is an usual way for all games. so it should be supported. […] so i would click one shortcut and iam ingame in a few seconds.

It is supported. See my explanation above for how to do it.


> just type /su foobar and then iam acting with his privs.

This might be possible with LUA mods. If not, feel free to open a feature request issue.

**Comment by codeandfix** [comment]
to clarify:
i think that gaming passwords arent worth a penny, except the admin password/griefing abilities.
so i dont care about password protection clientside as i think the local machine can always be fully trusted if the password is stored locally in plaintext. no need to technical stuff. just good old save password to ini times are perfect for a simple game ;-)

**Comment by mk-pmb** [comment]
If we start adding a simple login manager, we inevitably invite feature creep:
* In my case, the obvious next feature would be to sync my logins between my computers.
* In scenarios where people have to share their computer (e.g. some family scenarios), they'll want a more secure storage.

We could repeatedly explain to them why we're making a partially-good solution. Or we could incentivise people to use one of the existing solutions right from day one.

**Comment by codeandfix** [comment]
for multi-user, just write the data to user config as os standard.
otherwise write to game directory in (i prefer this) and select wether to store credentials or not.
ideally look for minetest.inf/conf inside minetest directory, if not found look into user directory or vice versus.
so both can work without much more coding effort.
otherwise double hash the password, so if local password is lost, just the hashed and salted pwd is lost. so the word is still secret.

**Comment by mk-pmb** [comment]
> for multi-user, just write the data to user config as os standard.

Won't be useful for average fanilies. They only have one user account on their machine, the one that came pre-installed.

Once we have proper login via URL handler, why should we maintain extra code to store passwords?

**Comment by codeandfix** [comment]
url handlers can be nice for private - not listed servers.
i think a command line parameter would be much better.
so users can add desktop shortcuts to their server. and if they dont like their family to access their account, they can input the passwort and/or playername manually.

i think support for one-user-accounts home user is not what minetest should solve.
just provide a simple credentials login at first time like cmd-line parameters (which is usual in nearly every good game/software).
if this is supported an url handler or more can come later.
i think a commandline for simple connect+user+pwd can be coded very quick with less effort.

**Comment by mk-pmb** [comment]
>  i think a command line parameter would be much better. […] so users can add desktop shortcuts to their server.

We do have the CLI params. You can already make direct-login desktop shortcuts. I already reminded you to look above. (Or search for `--password` in this thread.) I hope you're not trolling.

**Comment by codeandfix** [comment]
interesting. tried it with parameters. i thought they are server only. but with --go (which was not very intuitive to use) it works.
now i can connect from startmenu/desktop shortcut without having to type anything. thats it.
for me thats enough. if i would play on other servers i would prefer storing plain text credentials along a favorite, too.

**Comment by Hawk777** [comment]
This still seems like something nice to have. While making multiple shortcuts with command line parameters is certainly a workaround, it seems like it’s somewhat at odds with the presence of favourites in the server list. The fact that you can favourite servers in the server list suggests that that’s a reasonable way to connect to your three or four favourite servers; the fact that the client doesn’t save passwords in turn encourages using the same password across the servers. You *can* make multiple shortcuts to launch MT with the proper server name and password, but the fact that there’s a favourites system IMO encourages people to use that system, which, at this point, in turn encourages password reuse.

**Comment by mk-pmb** [comment]
> it’s somewhat at odds with the presence of favourites

I agree. I'd prefer to have this, and also the server list, as a separate project. The default download for shallow-interested users should bundle MT with our favorite server manager, accompanied by a slightly less visible download for just MT.

**Comment by Hawk777** [comment]
> I'd prefer to have the server list as a separate project.

Maybe that makes sense, but would you also suggest that managing singleplayer save files belongs in a separate tool? They’re conceptually the same thing (deciding where you want to play), yet even for people who are happy to make a separate shortcut for each server they play on, would they also be happy to make a separate shortcut for every save file? Or alternatively, if you bundle singleplayer save file management but not server management into the engine, then it looks really weird that you can start a singleplayer game but not a multiplayer game.

**Comment by mk-pmb** [comment]
Server management includes the complexity of credentials and logins, so yes, that deserves a separate project.

For single player worlds, I don't have the knowledge to decide whether it's complex enough. I could imagine lots of cool features that might be involved in a single player world, like once we get proper CSM inter-process communication, launching the external programs for some of the communicating mods in that world, or managing their credentials for external databases and stuff. At that point, a single player world could have enough aspects unrelated to the core MT engine, that it would make a lot of sense to offer a choice in single player world managers.

Thinking of it, don't we currently spawn a server for single player mode? And that server's server-side mods should be able to have stuff like external database connections with credentials, right? Or capability to write log files, which maybe should be prepared in some way, e.g. if the log file happens to be a serial modem line. :-D

**Comment by mk-pmb** [comment]
Also the fact that I need 3rd party tools to help me select which mods shall be enabled in my single player worlds, probably indicates that they have sufficient complexity already.

**Comment by codeandfix** [comment]
also i suggest to add a device id to minetest.
this id should be derivated like ms copy protection from things like c: hdd serial or bios serial or mac address etc.
so most griefers could be banned and a reinstall wont help.
only very advised users could circumvent this.
griefing is one of the biggest problem in minetest.
a hidden id which is stored on hdd wont help. but an derived id would be good.
problem is, that the derivation algorithm has to be a function which is bijective and unique.
like: use the first #0 network adapter's HW mac addess as hardware id.
so if i reinstall minetest on the same machine, the id has to be the same, even if all data has been destroyed.


another thought:
also this could be used as second factor of authentification.
so we could have low secure saved passwords if we force the same hardware id (hwid).
if hwid is not the same, then we should consider thinking about sending an email to the account if present (like a big game launcher does).
main problem here is abuse of server mailsystem for DoS "Spam".
also it would be nice to force players register with an email first. 
so we could have a blacklist for anti spam mailservices etc.

**Comment by MisterE123** [comment]
I would like this feature. 
I opened a new issue, but am closing it now that Im aware of this one. Here are my thoughts:

I keep typing my Minetest password into the wrong window (into the IRC window lol) and subsequently having to change it when Im trying to join my Minetest server
Especially when my computer slows down, sometimes I think Im typing my password into minetest's join game window, but im instead typing it into IRC. Which is bad lol.

Its also annoying to try to remember different Usernames and Passwords across different servers. For example, I have forgotten my Blocky Survival username and password.


Solutions

make a password manager, that stores your Username and password of each server. If you use multiple accounts on the server, present a dropdown to select the account you wish to use. When you join a server for the first time, it asks you if you want to save the account info (a checkbox, unchecked by default), because this could be a "public" minetest client (like at a library or a shared family computer where you dont want your siblings logging in as you).

also, there should be the possibility to manually add account info per server, and also to automatically join as a certain account to any new server. Also the possibility to export your user info as a .zip and import it to another client.

There should also be a warning to not use a password for another service.

This feature should be strictly opt-in.

**Comment by mk-pmb** [comment]
@MisterE123 Thanks for supporting this feature request!
Most of your ideas, as far as I understand them, have been discussed above already, some with better alternatives. When you've read up, please help future readers save time by removing mere repetitions from your post, or in cases of disagreement, clarify your reasons.

**Comment by SwissalpS** [comment]
In the meantime you can make per server scripts that log you in using the command line arguments and --go tag.
See: https://github.com/minetest/minetest/issues/12360#issuecomment-1134489075

**Comment by ProgrammerWhoPrograms** [comment]
[SQRL](https://www.grc.com/sqrl/sqrl.htm) is also an option if security is a concern. It would also remove the need to remember passwords.
Logging in to the same server on different machines might be difficult though? (same with token / session ID)

**Comment by beepbopbeepboop** [comment]
Would be nice to identify an open source, respected authentication library, and use it.  A good library is respected, trusted, open source, and already does everything everybody wants.  I was looking for a passkey library, seems someone forgot to create one. :-( But any solution is better than a user expecting their password to be saved, play for  10 hours, come back the next day, and viola, they will never be able to log in as they don't recall their password.

**Comment by esijg** [comment]
Just want to add to this conversation that for my family this feature is an accessability feature as my daughter is unable to join our servers on her own as she does not have the mental capacity (yet) to understand the concept of passwords, enter it, and log in. Although my daughter will eventually learn this in the coming months, I could see this being an accessibility problem for people with certain disabilities.

I personally have memory/cognitive problems that actually mirror early onset alzheimer symptoms (although thankfully reversible, but will take time) but thankfully I do have the technical knowledge and organizational skills to keep all my passwords neatly organized in a password manager. (which is fine until I forget about the existance of my password manager... which has happened but thankfully it's rare)

**Comment by mk-pmb** [comment]
The description of your challenges sounds like you would profit even more from automating the entire login process, including server selection. [These are the relevant parts of my launcher](https://github.com/mk-pmb/minetest-util-pmb/blob/8278220/launcher/src/decide_minetest_invocation.sh#L20-L29).
If you have really low security requirements it might even be as simple as making a desktop shortcut for MineTest, editing that shortcut and then adding ` --address my-server.local --port 12345 --name MyName --password MyPassword --go` to the command.
If none of this fits, feel free to make a feature request issue at my launcher repo to adapt parts of it to your software environment and circumstances. (The better you describe them, the better I can probably help.)

**Comment by MXB2001** [comment]
1986 is calling and wants its write passwords down on paper method back. Seriously, how does Luanti not have the ability to store usernames/passwords for each server? Or wait, is Luanti actually really from 1986? Must be...

**Comment by alwayshopeless** [comment]
Please, please, please, please 😿

**Comment by fgaz** [comment]
I looked a bit more into cross-platform keychain libraries while revisiting #12943, and I found another possible candidate: https://github.com/hrantzsch/keychain

It's maintained, it uses the system's keychain, it's written in C++, and it doesn't require any problematic dependency.

**Comment by appgurueu** [comment]
Today I learned that apparently Firefox's built-in password manager basically saves passwords in obfuscated, easily reversible form by default (if you don't set a master password). It doesn't even use the system keychain. (Chrome apparently does.)

So I figure maybe we've been overestimating the expected level of protection.

**Comment by alwayshopeless** [comment]
This discussion makes it sound like Minetest is the primary target of stealers worldwide.
Perhaps, for complete security, we should develop hardware password storage and only then add a login and password storage feature to the Luanti client?

**Comment by mk-pmb** [comment]
> we should develop hardware password storage

There are USB dongles that you can program to act like a keyboard and type pre-defined texts, like your login.
For most casual players, the desktop shortcut solution is probably safe enough.
Edit: I also made a new issue for Shibboleth. Most ideas for paranoia enhancements could probably facilitate or even reuse that.

**Comment by fgaz** [comment]
> This discussion makes it sound like Minetest is the primary target of stealers worldwide.
Perhaps, for complete security, we should develop hardware password storage and only then add a login and password storage feature to the Luanti client?

Given that libraries like [the one I linked above](https://github.com/luanti-org/luanti/issues/1566#issuecomment-3993713722) exist, using the system's keychain is only marginally more difficult than storing plain-text passwords.

Actually I wonder how much time it would take to add it to #16294. I might give it a try.

**Comment by MXB2001** [comment]
The passwords would be safely inside my PC. I wonder why all the angst about saving passwords...

**Comment by fgaz** [comment]
> Actually I wonder how much time it would take to add it to #16294. I might give it a try.

It indeed didn't take much time. Here's a proof of concept: #17162. I'd like to get core dev support before continuing with the to-do list.

**Comment by AliceDTRH** [comment]
This feature would be highly appreciated. I play on Steamdeck and having to type in a long password every time I want to join is kinda of a pain. It doesn't even remember it between multiple joins at the moment.
